flink-clients/src/test/java/org/apache/flink/client/program/PackagedProgramRetrieverImplTest.java

```diff
@@ -0,0 +1,573 @@
+            expectedClasspath.add(FileUtils.toURL(relativePath).toString());
+        }
+
+        assertThat(
+                actualClasspath,
+                IsIterableContainingInAnyOrder.containsInAnyOrder(expectedClasspath.toArray()));
+    }
+
+    @Test
+    public void testRetrieveFromJarFileWithoutUserLib()
+            throws IOException, FlinkException, ProgramInvocationException {
+        final PackagedProgramRetriever retrieverUnderTest =
+                PackagedProgramRetrieverImpl.create(
+                        null,
+                        classpathProvider.getTestJobJar(),
+                        null,
+                        classpathProvider.getTestJobArgs("suffix"),
+                        new Configuration());
+        final JobGraph jobGraph = retrieveJobGraph(retrieverUnderTest, new Configuration());
+
+        assertThat(
+                jobGraph.getUserJars(),
+                containsInAnyOrder(
+                        new org.apache.flink.core.fs.Path(
+                                classpathProvider.getTestJobJar().toURI())));
+        assertThat(jobGraph.getClasspaths().isEmpty(), is(true));
+    }
+
+    @Test
+    public void testRetrieveFromJarFileWithUserLib()
+            throws IOException, FlinkException, ProgramInvocationException {
+        final PackagedProgramRetriever retrieverUnderTest =
+                PackagedProgramRetrieverImpl.create(
+                        classpathProvider.getUserDirectoryWithOneEntryClass(),
+                        classpathProvider.getTestJobJar(),
+                        null,
+                        classpathProvider.getTestJobArgs("suffix"),
+                        new Configuration());
+        final JobGraph jobGraph = retrieveJobGraph(retrieverUnderTest, new Configuration());
+
+        assertThat(
+                jobGraph.getUserJars(),
+                containsInAnyOrder(
+                        new org.apache.flink.core.fs.Path(
+                                classpathProvider.getTestJobJar().toURI())));
+        final List<String> actualClasspath =
+                jobGraph.getClasspaths().stream().map(URL::toString).collect(Collectors.toList());
+        final Path workingDirectory = FileUtils.getCurrentWorkingDirectory();
+        final List<String> expectedClasspath = new ArrayList<>();
+        for (File file : classpathProvider.getUserDirectoryWithOneEntryClass().listFiles()) {
+            if (!file.getName().endsWith("jar")) {
+                // only jars are expected
+                continue;
+            }
+
+            Path relativePath = FileUtils.relativizePath(workingDirectory, file.toPath());
+            expectedClasspath.add(FileUtils.toURL(relativePath).toString());
+        }
+
+        assertThat(
+                actualClasspath,
+                IsIterableContainingInAnyOrder.containsInAnyOrder(expectedClasspath.toArray()));
+    }
+
+    @Test
+    public void testChildFirstDefaultConfiguration() throws FlinkException {
```

Comment:
both this test and the one below should be renamed to clarify that they are about the class loaders

Reply:
Their actual purpose is to check whether the configuration is acknowledged. The classloader is only the vehicle to test that. That's why, I would object renaming them. Or do you have some proposal for a better name? 🤔

Reply:
Shouldn't this then belong into a `PackagedProgramTest`?

Reply:
The actual bug which we want to fix in `FLINK-21445` is that we missed setting the configuration in the `ClassPathPackagedProgramRetriever` which is now replaced by `PackagedProgramRetrieverImpl`. That's why I added the tests here: To cover this behavior